pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1543 @@
+
+      // each new partition should have exactly 1 new segment in CONSUMING state, and metadata in IN_PROGRESS state
+      for (Integer partitionId : newPartitions) {
+        List<String> allSegmentsForPartition = partitionToAllSegmentsMap.get(partitionId);
+        if (!skipNewPartitions && idealState.isEnabled()) {
+          Assert.assertEquals(allSegmentsForPartition.size(), 1);
+          for (String segment : allSegmentsForPartition) {
+            Map<String, String> instanceStateMap = updatedMapFields.get(segment);
+            Assert.assertEquals(instanceStateMap.size(), nReplicas);
+            for (Map.Entry<String, String> entry : instanceStateMap.entrySet()) {
+              Assert.assertEquals(entry.getValue(), "CONSUMING");
+            }
+            LLCRealtimeSegmentZKMetadata newPartitionMetadata = updatedMetadataMap.get(segment);
+            Assert.assertNotNull(newPartitionMetadata);
+            Assert.assertEquals(newPartitionMetadata.getStatus(), CommonConstants.Segment.Realtime.Status.IN_PROGRESS);
+          }
+        } else {
+          Assert.assertEquals(allSegmentsForPartition.size(), 0);
+        }
+      }
+    } catch (Exception e) {
+      Assert.assertTrue(exceptionExpected);
+    }
+  }
+
+  private void testValidateLLCRepair() throws InvalidConfigException {
+
+    FakePinotLLCRealtimeSegmentManager segmentManager = new FakePinotLLCRealtimeSegmentManager(null);
+    String tableName = "repairThisTable_REALTIME";
+    int nReplicas = 2;
+    IdealStateBuilderUtil idealStateBuilder = new IdealStateBuilderUtil(tableName);
+
+    TableConfig tableConfig;
+    IdealState idealState;
+    int nPartitions;
+    List<String> instances = getInstanceList(5);
+
+    tableConfig = makeTableConfig(tableName, nReplicas, KAFKA_OFFSET, DUMMY_HOST, DEFAULT_SERVER_TENANT,
+        DEFAULT_STREAM_ASSIGNMENT_STRATEGY);
+    nPartitions = 4;
+    segmentManager._partitionAssignmentGenerator.setConsumingInstances(instances);
+    PartitionAssignment expectedPartitionAssignment =
+        segmentManager._partitionAssignmentGenerator.generatePartitionAssignment(tableConfig, nPartitions);
+
+    // set up a happy path - segment is present in ideal state, is CONSUMING, metadata says IN_PROGRESS
+    idealState = clearAndSetupHappyPathIdealState(idealStateBuilder, segmentManager, tableConfig, nPartitions);
+
+    // randomly introduce an error condition and assert that we repair it
+    long seed = new Random().nextLong();
+    System.out.println("Random seed for validate llc repairs " + seed);
+    Random random = new Random(seed);
+
+    for (int run = 0; run < 200; run++) {
+
+      boolean consuming = true;
+      boolean inProgress = true;
+      if (random.nextBoolean()) {
+        segmentManager.tooSoonToCorrect = true;
+      }
+      if (!random.nextBoolean()) { // not present in ideal state
+        // 1. very first metadata IN_PROGRESS, but segment missing - repair: create segment CONSUMING
```

Comment:
the comments are not clear. Would it help if you create a few final boolean variables for each case?
1. allSegmentsOffline
2. step3NotDone
3. tooSoon
4. step2NotDone
5. instancesChanged
6. nPartitionsIncreased

and then run the individual tests (perhaps in different methods if necessary)